examples/cis3-nft-sponsored-txs/src/lib.rs

```diff
@@ -0,0 +1,2083 @@
+    let old_public_key =
+        host.state_mut().public_key_registry.insert(params.account, params.public_key);
+
+    // Return an error if the owner tries to update/re-write a new public key to an
+    // already registered account.
+    ensure!(old_public_key.is_none(), CustomContractError::AccountAlreadyRegistered.into());
+
+    // Log the registration event.
+    logger.log(&Event::Registration(RegistrationEvent {
+        account:    params.account,
+        public_key: params.public_key,
+    }))?;
+
+    Ok(())
+}
+
+/// Mint new tokens with a given address as the owner of these tokens.
+/// Can only be called by the contract owner.
+/// Logs a `Mint` and a `TokenMetadata` event for each token.
+/// The url for the token metadata is the token ID encoded in hex, appended on
+/// the `TOKEN_METADATA_BASE_URL`.
+///
+/// It rejects if:
+/// - The sender is not the contract instance owner.
+/// - Fails to parse parameter.
+/// - Any of the tokens fails to be minted, which could be if:
+///     - The minted token ID already exists.
+///     - Fails to log Mint event
+///     - Fails to log TokenMetadata event
+///
+/// Note: Can at most mint 32 token types in one call due to the limit on the
+/// number of logs a smart contract can produce on each function call.
+#[receive(
+    contract = "cis3_nft",
+    name = "mint",
+    parameter = "MintParams",
+    error = "ContractError",
+    enable_logger,
+    mutable
+)]
+fn contract_mint<S: HasStateApi>(
+    ctx: &impl HasReceiveContext,
+    host: &mut impl HasHost<State<S>, StateApiType = S>,
+    logger: &mut impl HasLogger,
+) -> ContractResult<()> {
+    // Get the contract owner
+    let owner = ctx.owner();
+    // Get the sender of the transaction
+    let sender = ctx.sender();
+
+    ensure!(sender.matches_account(&owner), ContractError::Unauthorized);
+
+    // Parse the parameter.
+    let params: MintParams = ctx.parameter_cursor().get()?;
+
+    let (state, builder) = host.state_and_builder();
+
+    for &token_id in params.tokens.iter() {
+        // Mint the token in the state.
+        state.mint(token_id, &params.owner, builder)?;
+
+        // Event for minted NFT.
+        logger.log(&Cis2Event::Mint(MintEvent {
+            token_id,
+            amount: ContractTokenAmount::from(1),
+            owner: params.owner,
+        }))?;
+
+        // Metadata URL for the NFT.
+        logger.log(&Cis2Event::TokenMetadata::<_, ContractTokenAmount>(TokenMetadataEvent {
+            token_id,
+            metadata_url: MetadataUrl {
+                url:  build_token_metadata_url(&token_id),
+                hash: None,
+            },
+        }))?;
+    }
+    Ok(())
+}
+
+type TransferParameter = TransferParams<ContractTokenId, ContractTokenAmount>;
+
+/// Internal `transfer/permitTransfer` helper function. Invokes the `transfer`
+/// function of the state. Logs a `Transfer` event and invokes a receive hook
+/// function.
+fn _transfer<S: HasStateApi>(
```

Comment:
We don't really use the `_` prefix for internal functions. It is internal since it doesn't have a `receive` attribute, or a `pub` keyword in front 😊